docs/auth/sexpr.md

```diff
@@ -0,0 +1,563 @@
+  permission_ well-known class. A call to a base permission returns a
+  JSON array `[perm ...arguments]`, but note the arguments have been
+  evaluated. (Scalar)
+
+* Templates are named by UUIDs which are members of the _Permission
+  template_ well-known class and have a _Permission template definition_ 
+  ConfigDB entry. Bindings are created for the template parameters from
+  the argument list, and the results list is list-evaluated using (only)
+  these bindings and returned. (List)
+
+* Other calls evaluate as though they were an appropriate call to the
+  `throw` builtin.
+
+Note that the name of a call is not evaluated and must be specified as a
+literal string. There is no dynamic-call facility. The language has no
+lambdas and no first-class functions so the loss of flexibility is
+minor, and the restriction enables useful static analysis.
+
+### Builtins
+
+Builtins are all named by short, lowercase strings. Builtins fall into
+two categories: those which evaluate and flatten their arguments as
+though they were normal template calls with outside sources of
+information, and 'special forms' which evaluate their arguments
+differently.
+
+Special forms are denoted (*) below; these must be explicitly in the
+form documented, before evaluation. The notes '(Scalar)' and '(List)'
+mean the same as above.
+
+* `equal`: `["equal" a b]`. This performs deep equality testing on two
+  JSON values. Returns `true` or `false`. (Scalar)
+
+* `flat`: `["flat" list]`. Given a list value which has not already been
+  flattened, e.g. from `list` or `quote` or `lookup`, return the list
+  such that it will flatten into a return list. (List)
+
+* `format`: `["format" str ...args]`. Performs sprintf formatting using
+  Node's `util.format`.
+
+* `has`: `["has" obj key]`. Expects the first argument to be an object
+  and the second to be a string. Returns `true` if the object has the
+  named key, `false` otherwise. (Scalar)
+
+* `id`: `["id" principal type]`. Looks up identity details for the given
+  principal. The type can be `kerberos` or `sparkplug`.
+
+* `if` (*): `["if" cond pass]` or `["if" cond pass fail]`. Both forms
+  start by evaluating the condition. If this results in the empty list,
+  or if the first item is `null` or `false`, we evaluate the `fail`
+  value, or return the empty list. Otherwise we evaluate the `pass`
+  value. (List)
+
+* `let` (*): `["let" bindings ...body]`. The `bindings` must be a list
+  with an even number of items; odd numbered items must be strings and
+  name new bindings. Each is followed by a value. A new set of bindings
+  is constructed starting from the current set. The first value is
+  evaluated using the current bindings and a binding for the first name
+  is added with the result. Each subsequent value is evaluated using the
+  new bindings and added to the set. Finally `body` is list-evaluated
+  using the new bindings and the result returned. (List)
```

Comment:
Can you explain this in different terms? I don't entirely understand and as such can't work out what line 294 is doing below.

Reply:
This is 'variable' assignment (scare quotes because they aren't actually variable). It's equivalent to a block in Javascript with a set of `const` assignments at the top. Given `[let [a VALUE] ...]` the `...` is evaluated, and anywhere inside we replace `[a]` with `VALUE`. So:
```
[let [a 42] [a]] -> [42]
[let [a 42 b "abc"] [b] [a]] -> ["abc" 42]
[let [a 42 b "abc"] [format "%s/%s" [a] [b]]] -> ["42/abc"]
```
In the example on L294, `node` is a variable created for the scope of the let. `[id [principal] sparkplug]` is the value assigned to that variable, which evaluates to an object like `{ group: "Core", node: "Node" }` giving our assigned Sparkplug address. The body of the let contains a map, which iterates over a list built using the `node` variable we assigned to; the map creates another variable `addr`, and the expansion below shows the various stages in evaluating it.

Reply:
I suppose given this is JSON we could actually use the syntax `[let {a: 42} [a]]`. I wonder if that would be less confusing?